core/network/src/consensus_gossip.rs

```diff
@@ -18,42 +18,57 @@
 //! Handles chain-specific and standard BFT messages.
 
 use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
 use std::time::{Instant, Duration};
 use log::{trace, debug};
 use futures::sync::mpsc;
 use rand::{self, seq::SliceRandom};
 use lru_cache::LruCache;
-use network_libp2p::NodeIndex;
+use network_libp2p::{Severity, NodeIndex};
 use runtime_primitives::traits::{Block as BlockT, Hash, HashFor};
 pub use crate::message::generic::{Message, ConsensusMessage};
 use crate::protocol::Context;
 use crate::config::Roles;
 
 // FIXME: Add additional spam/DoS attack protection: https://github.com/paritytech/substrate/issues/1115
 const MESSAGE_LIFETIME: Duration = Duration::from_secs(120);
-const DEAD_TOPICS_CACHE_SIZE: usize = 4096;
+const KNOWN_MESSAGES_CACHE_SIZE: usize = 4096;
 
 struct PeerConsensus<H> {
 	known_messages: HashSet<H>,
 	is_authority: bool,
 }
 
 struct MessageEntry<B: BlockT> {
-	topic: B::Hash,
 	message_hash: B::Hash,
+	topic: B::Hash,
 	message: ConsensusMessage,
-	broadcast: bool,
+	timestamp: Instant,
+}
+
+/// Message validation result.
+pub enum ValidationResult<H> {
+	/// Message is valid with this topic.
+	Valid(H),
+	/// Invalid message.
+	Invalid,
+	/// Obsolete message.
+	Expired,
+}
+
+/// Validates consensus messages.
+pub trait Validator<H> {
+	/// Validate consensus message.
+	fn validate(&self, kind: u32, data: &[u8]) -> ValidationResult<H>;
 }
 
 /// Consensus network protocol handler. Manages statements and candidate requests.
 pub struct ConsensusGossip<B: BlockT> {
-	peers: HashMap<NodeIndex, PeerConsensus<(B::Hash, B::Hash)>>,
-	live_message_sinks: HashMap<B::Hash, Vec<mpsc::UnboundedSender<ConsensusMessage>>>,
+	peers: HashMap<NodeIndex, PeerConsensus<B::Hash>>,
+	live_message_sinks: HashMap<B::Hash, Vec<mpsc::UnboundedSender<Vec<u8>>>>,
 	messages: Vec<MessageEntry<B>>,
-	known_messages: HashSet<(B::Hash, B::Hash)>,
-	known_dead_topics: LruCache<B::Hash, ()>,
-	message_times: HashMap<(B::Hash, B::Hash), Instant>,
-	session_start: Option<B::Hash>,
+	known_messages: LruCache<B::Hash, ()>,
+	validators: HashMap<u32, Arc<Validator<B::Hash>>>,
```

Comment:
The `kind` field is used to identify message structure. And it is not GRANDPA-specific in any way. It just tells which validator should validate the message. Having each validator to try and decode a message to tell if it is valid or not is just wasting resources and may lead to potential conflict in case there are multiple types of messages with the same `Codec` structure.